pkg/provisioningv2/rke2/planner/initnode.go

```diff
@@ -0,0 +1,241 @@
+	// If the current init node has an empty joinURL annotation, we can look to see if there are other init nodes that are more suitable
+	if initNodeFound {
+		// if the init node was found but doesn't have a joinURL, let's see if there is possible a more suitable init node.
+		possibleInitNodes := collect(plan, canBeInitNode)
+		for _, entry := range possibleInitNodes {
+			if entry.Machine.Annotations[JoinURLAnnotation] != "" {
+				// if a non-blank JoinURL was found, return that we found an init node but with an error
+				return true, "", fmt.Errorf("non-populated init node found, but more suitable alternative is available")
+			}
+		}
+		// if we got through all possibleInitNodes (or there weren't any other possible init nodes), return true that we found an init node with no error.
+		logrus.Debugf("rkecluster %s/%s: init node with empty JoinURLAnnotation was found, no suitable alternatives exist", rkeControlPlane.Namespace, rkeControlPlane.Spec.ClusterName)
+		return true, "", nil
+	}
+
+	return false, "", fmt.Errorf("init node not found")
+}
+
+// electInitNode returns a joinURL and error (if one exists) of an init node. It will first search to see if an init node exists
+// (using findInitNode), then will perform a re-election of the most suitable init node (one with a joinURL) and fall back to simply
+// electing the first possible init node if no fully populated init node is found.
+func (p *Planner) electInitNode(rkeControlPlane *rkev1.RKEControlPlane, plan *plan.Plan) (string, error) {
+	logrus.Debugf("rkecluster %s/%s: determining if election of init node is necessary", rkeControlPlane.Namespace, rkeControlPlane.Spec.ClusterName)
+	initNodeFound, joinURL, err := p.findInitNode(rkeControlPlane, plan)
+	if (initNodeFound && err == nil) || errors.Is(err, generic.ErrSkip) {
+		logrus.Debugf("rkecluster %s/%s: init node was already elected and found with joinURL: %s", rkeControlPlane.Namespace, rkeControlPlane.Spec.ClusterName, joinURL)
+		return joinURL, err
+	}
+	// If the joinURL (or an errSkip) was not found, re-elect the init node.
+	logrus.Debugf("rkecluster %s/%s: performing election of init node", rkeControlPlane.Namespace, rkeControlPlane.Spec.ClusterName)
+
+	possibleInitNodes := collect(plan, canBeInitNode)
+	if len(possibleInitNodes) == 0 {
+		logrus.Debugf("rkecluster %s/%s: no possible init nodes exist", rkeControlPlane.Namespace, rkeControlPlane.Spec.ClusterName)
+		return joinURL, generic.ErrSkip
+	}
+
+	// clear all etcd init node marks because we are re-electing our init node
+	etcdEntries := collect(plan, isEtcd)
+	for _, entry := range etcdEntries {
+		// Ignore all etcd nodes that are not init nodes
+		if !isInitNode(entry.Machine) {
+			continue
+		}
+		logrus.Debugf("rkecluster %s/%s: clearing init node mark on machine %s", rkeControlPlane.Namespace, rkeControlPlane.Spec.ClusterName, entry.Machine.Name)
+		if err := p.clearInitNodeMark(entry.Machine); err != nil && !errors.Is(err, generic.ErrSkip) {
```

Comment:
My reasoning for not returning the `ErrSkip` at this point is because I want to have all init nodes cleared before we proceed to re-election. Let's say we have two init nodes (multiple), a proper re-election won't be done if we return the moment we clear the init node mark on the first of two init nodes, because suddenly that second init node is the init node.

Reply:
that is a good idea. 

let me see if I can boil up a quick "cachesInvalidated" thing

Reply:
implemented